src/System.Data.SqlClient/src/System/Data/SqlClient/SqlBuffer.cs

```diff
@@ -1127,9 +1083,9 @@ private static void FillInTimeInfo(ref TimeInfo timeInfo, byte[] timeBytes, int
             timeInfo.scale = scale;
         }
 
-        private static int GetDateFromByteArray(byte[] buf, int offset)
+        private static int GetDateFromByteArray(ReadOnlySpan<byte> buf)
         {
-            return buf[offset] + (buf[offset + 1] << 8) + (buf[offset + 2] << 16);
+            return buf[0] + (buf[1] << 8) + (buf[2] << 16);
```

Comment:
Current implementation is 20 assembly instructions.

Alternative implementations:
1) **Better (18 instructions)**
```C#
private static int GetDateFromByteArray(ReadOnlySpan<byte> buf)
{
    ushort first2Bytes = MemoryMarshal.Read<ushort>(buf);
    return first2Bytes + (buf[2] << 16);
}
```

2) **Best (16 instructions)**
```C#
private static int GetDateFromByteArray(ReadOnlySpan<byte> buf)
{
    byte thirdByte = buf[2]; // Hoist most of the bounds checks on buf.
    return buf[0] + (buf[1] << 8) + (thirdByte << 16);
}
```

cc @AndyAyersMS